### ISSUE TITLE
BaseSettings descendants call default_factory on import

### ISSUE THREAD

**Issue by pomo-mondreganto** [samuelcolvin/pydantic#2515]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

When importing `BaseSettings` descendant from another file, `default_factory` functors are always called for uninstantiated models. See the code snippet attached below. `def factory()` is called once on import, even though `B` is not even instantiated. 

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: True
                 install path: /usr/local/anaconda3/envs/dirty_3.9/lib/python3.9/site-packages/pydantic
               python version: 3.9.2 | packaged by conda-forge | (default, Feb 21 2021, 05:02:20)  [Clang 11.0.1 ]
                     platform: macOS-11.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

The first file (`kek.py`):

```py
from pydantic import BaseSettings, Field

v = 0

def factory():
    print('factory called')
    global v
    v += 1
    return v


class A(BaseSettings):
    pass


class B(A):
    id: int = Field(default_factory=factory)

        
class C(A):
    pass
```

The second file:

```py
from kek import C
```

Expected output: empty.

Actual output: `factory called`.

It's also worth mentioning that `BaseModel` does not have the described behavior, importing the `BaseModel` descendants doesn't call `default_factory` functors.

**Comment by MihanixA** [comment]
I co-discovered this issue. 100% True

**Comment by PrettyWood** [comment]
Hi and thanks for reporting
`BaseSettings` inherits from `BaseModel` with other defaults for the config including `validate_all` which is set to True by default.
As explained in the [warning section in the doc](https://pydantic-docs.helpmanual.io/usage/models/#field-with-dynamic-default-value), to validate the default value of a field, we actually need to call the `default_factory`.
I guess you could find a workaround in your default factory or set `validate_all = False`. We could also allow people to not validate values with `default_factory`
Hope it helps in the meantime

**Comment by pomo-mondreganto** [comment]
@PrettyWood 

Thank you, that works for now. It's still kind of strange though, I believe it should be possible to validate all values (including the default ones) on initialization. Might be better to hide this functionality under some option (`validate_on_init`), too.

**Comment by PrettyWood** [comment]
@pomo-mondreganto you're absolutely right! I don't even remember why I did this 🤷 I opened a PR to remove this side effect! Thanks for reporting

**Comment by guhcampos** [comment]
I had also missed the warning and had my expectations incorrect.

I would suggest a lazy validation for all fields with a default factory is the reasonably expected way to go. If we are defining a factory it means we will only have the actual data in runtime at some point.

In my case I stumbled into this while writing unit tests: I kept getting validation errors in tests at points in code I was not instantiating a settings object, which got me quite confused until I found out it was being created at import time. I think most of us will agree that running code with side effects in import time is a bad thing.
